// @ionic/angular/dist/directives/navigation/stack-controller.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var StackController = /** @class */ (function () {
    function StackController(stack, containerEl, router, navCtrl, zone) {
        this.stack = stack;
        this.containerEl = containerEl;
        this.router = router;
        this.navCtrl = navCtrl;
        this.zone = zone;
        this.viewsSnapshot = [];
        this.views = [];
        this.skipTransition = false;
    }
    StackController.prototype.createView = function (enteringRef, route) {
        return {
            ref: enteringRef,
            element: (enteringRef && enteringRef.location && enteringRef.location.nativeElement),
            url: this.getUrl(route)
        };
    };
    StackController.prototype.getExistingView = function (activatedRoute) {
        var activatedUrlKey = this.getUrl(activatedRoute);
        return this.views.find(function (vw) { return vw.url === activatedUrlKey; });
    };
    StackController.prototype.setActive = function (enteringView, direction, animated) {
        return __awaiter(this, void 0, void 0, function () {
            var leavingView;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        leavingView = this.getActive();
                        this.insertView(enteringView, direction);
                        return [4 /*yield*/, this.transition(enteringView, leavingView, direction, animated, this.canGoBack(1), false)];
                    case 1:
                        _a.sent();
                        this.cleanup();
                        return [2 /*return*/];
                }
            });
        });
    };
    StackController.prototype.canGoBack = function (deep) {
        return this.views.length > deep;
    };
    StackController.prototype.pop = function (deep) {
        var _this = this;
        this.zone.run(function () {
            var view = _this.views[_this.views.length - deep - 1];
            _this.navCtrl.navigateBack(view.url);
        });
    };
    StackController.prototype.startBackTransition = function () {
        this.transition(this.views[this.views.length - 2], // entering view
        this.views[this.views.length - 1], // leaving view
        'back', true, true, true);
    };
    StackController.prototype.endBackTransition = function (shouldComplete) {
        if (shouldComplete) {
            this.skipTransition = true;
            this.pop(1);
        }
    };
    StackController.prototype.insertView = function (enteringView, direction) {
        // no stack
        if (!this.stack) {
            this.views = [enteringView];
            return;
        }
        // stack setRoot
        if (direction === 'root') {
            this.views = [enteringView];
            return;
        }
        // stack
        var index = this.views.indexOf(enteringView);
        if (index >= 0) {
            this.views = this.views.slice(0, index + 1);
        }
        else {
            if (direction === 'forward') {
                this.views.push(enteringView);
            }
            else {
                this.views = [enteringView];
            }
        }
    };
    StackController.prototype.cleanup = function () {
        var views = this.views;
        this.viewsSnapshot
            .filter(function (view) { return !views.includes(view); })
            .forEach(function (view) { return destroyView(view); });
        for (var i = 0; i < views.length - 1; i++) {
            var view = views[i];
            var element = view.element;
            element.setAttribute('aria-hidden', 'true');
            element.classList.add('ion-page-hidden');
            // view.ref.changeDetectorRef.detach();
        }
        this.viewsSnapshot = views.slice();
    };
    StackController.prototype.getActive = function () {
        var views = this.views;
        return views.length > 0 ? views[views.length - 1] : undefined;
    };
    StackController.prototype.transition = function (enteringView, leavingView, direction, animated, showGoBack, progressAnimation) {
        return __awaiter(this, void 0, void 0, function () {
            var enteringEl, leavingEl, containerEl;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.runningTransition) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.runningTransition];
                    case 1:
                        _a.sent();
                        this.runningTransition = undefined;
                        _a.label = 2;
                    case 2:
                        if (this.skipTransition) {
                            this.skipTransition = false;
                            return [2 /*return*/];
                        }
                        enteringEl = enteringView ? enteringView.element : undefined;
                        leavingEl = leavingView ? leavingView.element : undefined;
                        containerEl = this.containerEl;
                        if (!(enteringEl && enteringEl !== leavingEl)) return [3 /*break*/, 5];
                        enteringEl.classList.add('ion-page', 'ion-page-invisible');
                        if (enteringEl.parentElement !== containerEl) {
                            containerEl.appendChild(enteringEl);
                        }
                        return [4 /*yield*/, containerEl.componentOnReady()];
                    case 3:
                        _a.sent();
                        this.runningTransition = containerEl.commit(enteringEl, leavingEl, {
                            duration: !animated ? 0 : undefined,
                            direction: direction === 'forward' ? 'forward' : 'back',
                            deepWait: true,
                            showGoBack: showGoBack,
                            progressAnimation: progressAnimation
                        });
                        return [4 /*yield*/, this.runningTransition];
                    case 4:
                        _a.sent();
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    StackController.prototype.getUrl = function (activatedRoute) {
        var urlTree = this.router.createUrlTree(['.'], { relativeTo: activatedRoute });
        return this.router.serializeUrl(urlTree);
    };
    return StackController;
}());
export { StackController };
function destroyView(view) {
    if (view) {
        // TODO lifecycle event
        view.ref.destroy();
    }
}
